packages/kbn-esql-ast/src/types.ts

```diff
@@ -67,11 +67,67 @@ export interface ESQLCommandMode extends ESQLAstBaseItem {
   type: 'mode';
 }
 
-export interface ESQLFunction extends ESQLAstBaseItem {
+/**
+ * We coalesce all function calls and expressions into a single "function"
+ * node type. This subtype is used to distinguish between different types
+ * of function calls and expressions.
+ *
+ * - `variadic-call` is a function call with any number of arguments: fn(a, b, c, ...)
+ * - `unary-expression` is a unary expression: -a, +a, NOT a, ...
+ * - `binary-expression` is a binary expression: a + b, a - b, a * b, ...
+ */
+export type FunctionSubtype =
+  | 'variadic-call' // fn(a, b, c, ...)
+  | 'unary-expression' // -a, +a, NOT a, ...
+  | 'postfix-unary-expression' // a IS NULL, a IS NOT NULL, ...
+  | 'binary-expression'; // a + b, a - b, a * b, ...
+
+export interface ESQLFunction<
+  Subtype extends FunctionSubtype = FunctionSubtype,
+  Name extends string = string
+> extends ESQLAstBaseItem<Name> {
   type: 'function';
+
+  /**
+   * Default is 'variadic-call'.
+   */
+  subtype?: Subtype;
+
+  args: ESQLAstItem[];
+}
+
+export interface ESQLFunctionCallExpression extends ESQLFunction<'variadic-call'> {
+  subtype: 'variadic-call';
   args: ESQLAstItem[];
 }
 
+export interface ESQLUnaryExpression extends ESQLFunction<'unary-expression'> {
+  subtype: 'unary-expression';
+  args: [ESQLAstItem];
+}
+
+export interface ESQLPostfixUnaryExpression extends ESQLFunction<'postfix-unary-expression'> {
+  subtype: 'postfix-unary-expression';
+  args: [ESQLAstItem];
+}
+
+export interface ESQLBinaryExpression
+  extends ESQLFunction<'binary-expression', BinaryExpressionOperator> {
+  subtype: 'binary-expression';
+  args: [ESQLAstItem, ESQLAstItem];
+}
+
+export type BinaryExpressionOperator =
+  | BinaryExpressionArithmeticOperator
+  | BinaryExpressionAssignmentOperator
+  | BinaryExpressionComparisonOperator
+  | BinaryExpressionRegexOperator;
+
+export type BinaryExpressionArithmeticOperator = '+' | '-' | '*' | '/' | '%';
+export type BinaryExpressionAssignmentOperator = '=';
+export type BinaryExpressionComparisonOperator = '==' | '=~' | '!=' | '<' | '<=' | '>' | '>=';
```

Comment:
I don't think `=~` is a valid operator

Reply:
I took it from our ANTLR grammar:

<img width="270" alt="image" src="https://github.com/user-attachments/assets/3293864f-706e-4e0a-b544-676f78fa3343">

But you are also right, I don't see it being used. It is either unused code or something for the future.

Reply:
I can remove it, or leave it, indeifferent.